ui/utility.py

```diff
@@ -9,17 +9,43 @@
 import re
 import logging
 import functools
+import tarfile
+import shutil
+from datetime import datetime
 log = logging.getLogger(__name__)
 
 class FormfeedConversionException(Exception): pass
 class LinefeedConversionException(Exception): pass
 
-def find_firmware(directory):
+def update_ui(config):
+    log.info("updating UI")
+    usb_dir = config.get('files', 'usb_dir')
+    ui_file = find_ui_update(usb_dir)
+
+    log.info("looking for new ui in %s" % usb_dir)
+    install_dir = config.get('files', 'install_dir')
+    install_dir = os.path.expanduser(install_dir)
+
+    archive_dir = config.get('files', 'archive_dir')
+    archive_dir = os.path.expanduser(archive_dir)
+
+    # archive old ui
+    archive_dir += datetime.now().strftime("%Y%m%d-%H%M%S-ui")
+    log.info("archiving %s to %s" % (install_dir, archive_dir))
+    shutil.move(install_dir, archive_dir)
```

Comment:
What do you mean by "new UI software"? Is that an explanation or are you asking me?

I think it's safer to copy into the archive, halt the UI and then replace it. Removing the working directory and all files of a running program makes me very nervous.